File: medium/433. Minimum Genetic Mutation/433. Minimum Genetic Mutation.go
package main

import "fmt"

func minMutation(startGene string, endGene string, bank []string) int {
	bankSet := make(map[string]bool)
	for _, gene := range bank {
		bankSet[gene] = true
	}

	if !bankSet[endGene] {
		return -1
	}

	queue := []string{startGene}
	visited := map[string]bool{startGene: true}
	steps := 0
	genes := []byte{'A', 'C', 'G', 'T'}

	for len(queue) > 0 {
		nextQueue := []string{}

		for _, current := range queue {
			if current == endGene {
				return steps
			}

			for i := 0; i < len(current); i++ {
				for _, g := range genes {
					if current[i] == g {
						continue
					}

					mutated := current[:i] + string(g) + current[i+1:]
					if bankSet[mutated] && !visited[mutated] {
						visited[mutated] = true
						nextQueue = append(nextQueue, mutated)
					}
				}
			}
		}

		queue = nextQueue
		steps++
	}

	return -1
}

func main() {
	fmt.Println(minMutation("AACCGGTT", "AACCGGTA", []string{"AACCGGTA"}))
	fmt.Println(minMutation("AACCGGTT", "AAACGGTA", []string{"AACCGGTA", "AACCGCTA", "AAACGGTA"}))
}

File: medium/433. Minimum Genetic Mutation/433. Minimum Genetic Mutation.py
from typing import List


class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        bank_set = {gene: True for gene in bank}

        if endGene not in bank_set:
            return -1

        queue: List[str] = [startGene]
        visited = {startGene: True}
        steps = 0
        genes = {"A", "C", "G", "T"}

        while len(queue) > 0:
            next_queue: List[str] = []

            for current in queue:
                if current == endGene:
                    return steps

                for i in range(len(current)):
                    for g in genes:
                        if current[i] == g:
                            continue

                        mutated = current[:i] + g + current[i + 1 :]
                        if (
                            mutated in bank_set
                            and bank_set[mutated]
                            and mutated not in visited
                        ):
                            visited[mutated] = True
                            next_queue.append(mutated)

            queue = next_queue
            steps += 1

        return -1


solution = Solution()
print(solution.minMutation(startGene="AACCGGTT", endGene="AACCGGTA", bank=["AACCGGTA"]))
print(
    solution.minMutation(
        startGene="AACCGGTT",
        endGene="AAACGGTA",
        bank=["AACCGGTA", "AACCGCTA", "AAACGGTA"],
    )
)
